File: src/main/java/org/cubewhy/celestial/utils/game/MinecraftData.java
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.utils.game;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import okhttp3.Response;
import org.cubewhy.celestial.utils.RequestUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

public final class MinecraftData {
    public static final URL versionManifest;

    static {
        try {
            versionManifest = new URL("https://launchermeta.mojang.com/mc/game/version_manifest_v2.json");
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        }
    }

    private MinecraftData() {
    }

    public static JsonObject manifest() throws IOException {
        try (Response response = RequestUtils.get(versionManifest).execute()) {
            assert response.body() != null;
            return JsonParser.parseString(response.body().string()).getAsJsonObject();
        }
    }

    /**
     * Get information of a Minecraft release
     *
     * @param version version id
     * @return version json
     */
    public static @Nullable JsonObject getVersion(String version, @NotNull JsonElement json) throws IOException {
        JsonArray versionsArray = json.getAsJsonObject().getAsJsonArray("versions");
        for (JsonElement element : versionsArray) {
            if (element.getAsJsonObject().get("id").getAsString().equals(version)) {
                String url = element.getAsJsonObject().get("url").getAsString();
                try (Response response = RequestUtils.get(url).execute()) {
                    assert response.body() != null;
                    return JsonParser.parseString(response.body().string()).getAsJsonObject();
                }
            }
        }
        return null;
    }

    /**
     * Get texture index (Minecraft)
     *
     * @param json json object from MinecraftData.getVersion
     * @return json of texture index
     * */
    public static JsonObject getTextureIndex(@NotNull JsonElement json) throws IOException{
        URL url = new URL(json.getAsJsonObject().getAsJsonObject("assetIndex").get("url").getAsString());
        try (Response response = RequestUtils.get(url).execute()) {
            assert response.body() != null;
            return JsonParser.parseString(response.body().string()).getAsJsonObject();
        }
    }
}

File: src/main/kotlin/org/cubewhy/celestial/utils/game/MinecraftData.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */
package org.cubewhy.celestial.utils.game

import kotlinx.serialization.Serializable
import org.cubewhy.celestial.JSON
import org.cubewhy.celestial.string
import org.cubewhy.celestial.utils.RequestUtils.get
import java.net.URL

object MinecraftData {
    private var versionManifest: URL = URL("https://launchermeta.mojang.com/mc/game/version_manifest_v2.json")
    var texture: URL = URL("https://resources.download.minecraft.net")


    fun manifest(): MinecraftManifest {
        get(versionManifest).execute().use { response ->
            return JSON.decodeFromString(response.string!!)
        }
    }

    /**
     * Get information of a Minecraft release
     *
     * @param version version id
     * @return version json
     */

    fun getVersion(version: String, manifest: MinecraftManifest): MinecraftArtifactInfo? {
        for (versionInfo in manifest.versions) {
            if (versionInfo.id == version) {
                val url = versionInfo.url
                get(url).execute().use { response ->
                    return JSON.decodeFromString(response.string!!)
                }
            }
        }
        return null
    }

    /**
     * Get texture index (Minecraft)
     *
     * @param info json object from MinecraftData.getVersion
     * @return json of texture index
     */

    fun getTextureIndex(info: MinecraftArtifactInfo): MinecraftResources {
        val url = URL(info.assetIndex.url)
        get(url).execute().use { response ->
            return JSON.decodeFromString(response.string!!)
        }
    }
}

@Serializable
data class MinecraftResources(
    val objects: Map<String, Resource>
) {
    @Serializable
    data class Resource(
        val hash: String,
        val size: Int
    )
}

@Serializable
data class MinecraftManifest(
    val versions: List<MinecraftVersionInfo>
) {
    @Serializable
    data class MinecraftVersionInfo(
        val id: String,
        val url: String
    )
}

@Serializable
data class MinecraftArtifactInfo(
    val assets: String,
    val assetIndex: AssetIndex
) {
    @Serializable
    data class AssetIndex(
        val id: String,
        val sha1: String,
        val url: String
    )
}
